core-schema/src/schema.rs

```diff
@@ -0,0 +1,585 @@
+//! # "#)?;
+//! assert_eq!(schema.features, vec![
+//!     Feature {
+//!         spec: Spec::new("https://lib.apollo.dev/core", "core", (0, 1)),
+//!         name: Cow::Borrowed("core"),
+//!         position: Pos { line: 3, column: 7 },
+//!     },
+//!     Feature {
+//!         spec: Spec::new("https://spec.example.com/A", "A", (1, 0)),
+//!         name: Cow::Borrowed("A"),
+//!         position: Pos { line: 4, column: 7 },
+//!     },
+//! ]);
+//! # Ok::<(), graphql_parser::ParseError>(())
+//! ```
+
+use std::{borrow::Cow, collections::HashMap};
+
+use graphql_parser::{
+    parse_schema,
+    schema::{Definition, Document, SchemaDefinition},
+    ParseError, Pos,
+};
+
+use thiserror::Error;
+
+use crate::{constants::CORE, spec, Feature, Find, Found, Implementations, Version};
+
+/// A Schema holds a parsed GraphQL schema document and the features in that document,
+/// along with any errors which occurred during validation.
+pub struct Schema<'a> {
+    /// the document AST, as parsed by graphql_parser
+    pub document: Document<'a>,
+
+    /// features in the document
+    pub features: Vec<Feature>,
+
+    /// validation errors
+    pub errors: Vec<SchemaError>,
+}
+
+impl<'a> Schema<'a> {
+    /// Parse a schema source string and return a validated `Schema` with all
+    /// `Request`s identified, or fail with a `graphql_parser::ParseError`.
+    ///
+    /// As indicated by the type, this fn will always return `Ok` if parsing
+    /// succeeds. Subsequent errors are reported in the `errors` field of the
+    /// returned schema.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use core_schema::{Schema, Feature, Spec, Version};
+    /// use graphql_parser::{ParseError, Pos};
+    /// use std::borrow::Cow;
+    ///
+    /// let schema = Schema::parse(r#"
+    ///     schema
+    ///       @core(feature: "https://lib.apollo.dev/core/v0.1")
+    ///       @core(feature: "https://spec.example.com/A/v1.2", as: "exampleA")
+    ///       @core(feature: "https://example.net/specB/v0.1")
+    ///     {
+    ///       query: Query
+    ///     }
+    /// "#)?;
+    ///
+    /// assert_eq!(schema.errors, vec![]);
+    /// assert_eq!(schema.features, vec![
+    ///    Feature {
+    ///        spec: Spec::new("https://lib.apollo.dev/core", "core", (0, 1)),
+    ///        name: Cow::Borrowed("core"),
+    ///        position: Pos { line: 3, column: 7 },
+    ///    },        
+    ///    Feature {
+    ///        spec: Spec::new("https://spec.example.com/A", "A", (1, 2)),
+    ///        name: Cow::Borrowed("exampleA"),
+    ///        position: Pos { line: 4, column: 7 },
+    ///    },
+    ///    Feature {
+    ///        spec: Spec::new("https://example.net/specB", "specB", (0, 1)),
+    ///        name: Cow::Borrowed("specB"),
+    ///        position: Pos { line: 5, column: 7 },
+    ///    },
+    /// ]);
+    /// # Ok::<(), ParseError>(())
+    /// ```
+    pub fn parse(source: &'a str) -> Result<Schema<'a>, ParseError> {
```

Comment:
Might it be a better API to take in a parsed schema as an argument rather than only working on strings, and then offering an alternate entry point that takes care of both parsing and core-ing?